### PR TITLE
Remove downcast constructor from generated C++ proxy classes.

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR removes the downcast constructor in generated C++ proxies, namely:

```cpp
explicit GreeterPrx(const Ice::ObjectPrx& other) // downcast constructor
```

This constructor was a more convenient alternative to `Ice::uncheckedCast<T>`.

This PR also adds template parameters to all proxy-returning functions:
 - Communicator::stringToProxy
 - Communicator::propertyToProxy (was already done)
 - ObjectAdapter::add, addFacet, addWithUUID, createProxy etc.
 - Connection::createProxy

A well-written Ice 3.8 C++ application should never create an ObjectPrx to immediately cast it to a typed Prx. It should always immediately create the typed Prx. 

As a consequence, `uncheckedCast` is a code-smell: it can be correct, but you should double-check if you really need it. And because it's a code smell, it should be more explicit/visible than an explicit conversion via a constructor.

It's like in Java and C#, you rarely create `java.lang.Object` / `object` to then cast it to your desired type. You should obtain directly your favorite type.

Note: unfortunately, I don't believe we can provide similar generic add / createProxy APIs in C#, Java, Swift (and more?) because proxies are not concrete types in these languages. So uncheckedCast will be remain very common in these languages.

### CODE REVIEW

File: cpp/src/Glacier2/Glacier2Router.cpp [hunk 1]
```diff
@@ -350,7 +350,7 @@ RouterService::start(int argc, char* argv[], int& status)
     // The session router is used directly as a servant for the main
     // Glacier2 router Ice object.
     //
-    Glacier2::RouterPrx routerPrx(clientAdapter->add(_sessionRouter, {"router", instanceName}));
+    Glacier2::RouterPrx routerPrx = clientAdapter->add<Glacier2::RouterPrx>(_sessionRouter, {"router", instanceName});
```
Comment (pepone) [inline cpp/src/Glacier2/Glacier2Router.cpp:353]: Maybe we should use `auto` to not repeat the type name in the declaration and the template parameter.

File: cpp/src/Ice/Instance.cpp [hunk 2]
```diff
@@ -674,7 +674,7 @@ IceInternal::Instance::setServerProcessProxy(const ObjectAdapterPtr& adminAdapte
     const string serverId = _initData.properties->getIceProperty("Ice.Admin.ServerId");
     if (locator && serverId != "")
     {
-        ProcessPrx process{admin->ice_facet("Process")};
+        ProcessPrx process = admin->ice_facet<ProcessPrx>("Process");
```
Comment (pepone) [inline cpp/src/Ice/Instance.cpp:677]: I would use `auto` here too.

File: cpp/src/Ice/Instance.cpp [hunk 3]
```diff
@@ -1521,7 +1521,7 @@ IceInternal::Instance::finishSetup(int& argc, const char* argv[], const Ice::Com
     //
     if (!_referenceFactory->getDefaultRouter())
     {
-        optional<RouterPrx> router{communicator->propertyToProxy("Ice.Default.Router")};
+        optional<RouterPrx> router{communicator->propertyToProxy<RouterPrx>("Ice.Default.Router")};
```
Comment (pepone) [inline cpp/src/Ice/Instance.cpp:1524]: Shouldn't we use asigment here? propertyToProxy already returns an optional<T>
```suggestion
        auto router = communicator->propertyToProxy<RouterPrx>("Ice.Default.Router");
```

File: cpp/src/Ice/Instance.cpp [hunk 4]
```diff
@@ -1530,7 +1530,7 @@ IceInternal::Instance::finishSetup(int& argc, const char* argv[], const Ice::Com
 
     if (!_referenceFactory->getDefaultLocator())
     {
-        optional<LocatorPrx> locator{communicator->propertyToProxy("Ice.Default.Locator")};
+        optional<LocatorPrx> locator{communicator->propertyToProxy<LocatorPrx>("Ice.Default.Locator")};
```
Comment (pepone) [inline cpp/src/Ice/Instance.cpp:1533]: ditto

File: cpp/src/IceBridge/IceBridge.cpp [hunk 5]
```diff
@@ -557,7 +557,7 @@ BridgeService::start(int argc, char* argv[], int& status)
     }
 
     string instanceName = properties->getIceProperty("IceBridge.InstanceName");
-    RouterPrx router(adapter->add(make_shared<RouterI>(), stringToIdentity(instanceName + "/router")));
+    RouterPrx router = adapter->add<RouterPrx>(make_shared<RouterI>(), stringToIdentity(instanceName + "/router"));
```
Comment (pepone) [inline cpp/src/IceBridge/IceBridge.cpp:560]:
```suggestion
    auto router = adapter->add<RouterPrx>(make_shared<RouterI>(), stringToIdentity(instanceName + "/router"));
```

File: cpp/src/IceDiscovery/PluginI.cpp [hunk 6]
```diff
@@ -115,7 +115,7 @@ PluginI::initialize()
     // Setup locator registry.
     //
     auto locatorRegistry = make_shared<LocatorRegistryI>(_communicator);
-    Ice::LocatorRegistryPrx locatorRegistryPrx(_locatorAdapter->addWithUUID(locatorRegistry));
+    Ice::LocatorRegistryPrx locatorRegistryPrx = _locatorAdapter->addWithUUID<Ice::LocatorRegistryPrx>(locatorRegistry);
```
Comment (pepone) [inline cpp/src/IceDiscovery/PluginI.cpp:118]:
```suggestion
    auto locatorRegistryPrx = _locatorAdapter->addWithUUID<Ice::LocatorRegistryPrx>(locatorRegistry);
```

File: cpp/src/IceGrid/IceGridNode.cpp [hunk 7]
```diff
@@ -440,7 +440,7 @@ NodeService::startImpl(int argc, char* argv[], int& status)
 
     // Create the server factory. The server factory creates persistent objects for the server and server adapter. It
     // also takes care of installing the evictors and object factories necessary to store these objects.
-    NodePrx nodeProxy{_adapter->createProxy(stringToIdentity(instanceName + "/Node-" + name))};
+    NodePrx nodeProxy = _adapter->createProxy<NodePrx>(stringToIdentity(instanceName + "/Node-" + name));
```
Comment (pepone) [inline cpp/src/IceGrid/IceGridNode.cpp:443]:
```suggestion
    auto nodeProxy = _adapter->createProxy<NodePrx>(stringToIdentity(instanceName + "/Node-" + name));
```

**Review by pepone — APPROVED** [review]
Looks good, we can use auto in a few more places with the new templates but it is ok to do in a follow-up PR.

File: cpp/src/Glacier2/Glacier2Router.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Glacier2/Glacier2Router.cpp:353]: Yes. I believe I've fixed them all.

**Pull request merged**